File: FromSlides/C/09_global_loop_critical.c
/*THIS PROGRAM DOESN'T WORK PROPERLY*/

#include <stdio.h>
#include <omp.h>

#define MAX_ITS 10000

int main(){

  int its_global, i;
  
  its_global = 0;
#pragma omp parallel for
  for (i=0;i<MAX_ITS;++i){
/*The critical section ensures that only one thread can be in it at a time
this makes the incrementing of its_global safe*/

#pragma omp critical
    {
      its_global++;
    }
  }

  printf("Counter records %i iterations\n", its_global);
}

File: FromSlides/Fortran/09_global_loop_critical.f90
PROGRAM loop_decompose
  
  USE omp_lib
  IMPLICIT NONE
  INTEGER, PARAMETER :: max_its = 10000
  INTEGER :: i, its_global


its_global = 0
!$OMP PARALLEL DO
  DO i = 1, max_its
!The critical section ensures that only one thread can be in it at a time
!this makes the incrementing of its_global safe
!$OMP CRITICAL
    its_global = its_global + 1
!$OMP END CRITICAL
  END DO
!$OMP END PARALLEL DO

PRINT '(A,I0,A)','Counter records ', its_global, ' iterations'

END PROGRAM loop_decompose
